core/src/appleMain/kotlin/CBPeripheralCoreBluetoothPeripheral.kt

```diff
@@ -164,35 +160,64 @@ internal class CBPeripheralCoreBluetoothPeripheral(
             logger.verbose { message = "Configuring characteristic observations" }
             observers.onConnected()
         } catch (e: Exception) {
-            logger.error(e) { message = "Failed to connect" }
-            withContext(NonCancellable) {
-                centralManager.cancelPeripheralConnection(cbPeripheral)
-            }
-            throw e
+            closeConnection()
+            val failure = e.unwrap()
+            logger.error(failure) { message = "Failed to connect" }
+            throw failure
         }
 
-        centralManager.delegate.onDisconnected.onEach { identifier ->
-            if (identifier == cbPeripheral.identifier) {
-                connectAction.reset()
-                logger.info { message = "Disconnected" }
-            }
-        }.launchIn(scope)
-
         logger.info { message = "Connected" }
         _state.value = State.Connected
+
+        centralManager.delegate.state.watchForDisablingIn(scope)
+        centralManager.delegate.onDisconnected.watchForConnectionLossIn(scope)
     }
 
-    override suspend fun disconnect() {
-        try {
-            connectAction.resetAndJoin()
-        } finally {
-            withContext(NonCancellable) {
-                centralManager.cancelPeripheralConnection(cbPeripheral)
+    private fun Flow<CBManagerState>.watchForDisablingIn(scope: CoroutineScope) =
+        filter { state -> state != CBManagerStatePoweredOn }
+            .onEach { state ->
+                logger.info {
+                    message = "Bluetooth unavailable"
+                    detail("state", state)
+                }
+                closeConnection()
+                throw ConnectionLostException("$this $state")
+            }
+            .launchIn(scope)
+
+    private fun Flow<NSUUID>.watchForConnectionLossIn(scope: CoroutineScope) =
+        filter { identifier -> identifier == cbPeripheral.identifier }
+            .onEach {
+                logger.info { message = "Disconnected" }
+                throw ConnectionLostException("$this disconnected")
             }
-            logger.info { message = "Disconnected" }
+            .launchIn(scope)
+
+    override suspend fun disconnect() {
+        closeConnection()
+        suspendUntil<Disconnected>()
+        logger.info { message = "Disconnected" }
+    }
+
+    private suspend fun closeConnection() {
+        withContext(NonCancellable) {
+            centralManager.cancelPeripheralConnection(cbPeripheral)
+        }
+    }
+
+    private fun dispose(cause: Throwable?) {
+        GlobalScope.launch(start = UNDISPATCHED) {
```

Comment:
`dispose` is called from an `invokeOnCompletion` (of the peripheral scope) so cannot be `suspend`.
Since everything is shutting down at that point, it seems fine to lean on `GlobalScope` to finish shutting down the connection. We need to run within a coroutine scope because `cancelPeripheralConnection` uses a dispatcher for executing the cancel peripheral connection request.